rust/headless-client/src/dns_control/windows.rs

```diff
@@ -82,49 +78,50 @@ pub(crate) fn system_resolvers() -> Result<Vec<IpAddr>> {
     Ok(resolvers)
 }
 
+// Our NRPT rule should always live at
+// `Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Dnscache\Parameters\DnsPolicyConfig\{6C0507CB-C884-4A78-BC55-0ACEE21227F6}`
+const NRPT_REG_KEY: &str = "{6C0507CB-C884-4A78-BC55-0ACEE21227F6}";
+
 /// Tells Windows to send all DNS queries to our sentinels
 ///
 /// Parameters:
 /// - `dns_config_string`: Comma-separated IP addresses of DNS servers, e.g. "1.1.1.1,8.8.8.8"
-pub(crate) fn activate(dns_config: &[IpAddr]) -> Result<()> {
-    let dns_config_string = dns_config
-        .iter()
-        .map(|ip| format!("\"{ip}\""))
-        .collect::<Vec<_>>()
-        .join(",");
-
-    // Set our DNS IP as the DNS server for our interface
+fn activate(dns_config: &[IpAddr]) -> Result<()> {
     // TODO: Known issue where web browsers will keep a connection open to a site,
     // using QUIC, HTTP/2, or even HTTP/1.1, and so they won't resolve the DNS
     // again unless you let that connection time out:
     // <https://github.com/firezone/firezone/issues/3113#issuecomment-1882096111>
-    Command::new("powershell")
-        .creation_flags(CREATE_NO_WINDOW)
-        .arg("-Command")
-        .arg(format!(
-            "Set-DnsClientServerAddress {TUNNEL_NAME} -ServerAddresses({dns_config_string})"
-        ))
-        .status()?;
 
     tracing::info!("Activating DNS control");
-    Command::new("powershell")
-        .creation_flags(CREATE_NO_WINDOW)
-        .args([
-            "-Command",
-            "Add-DnsClientNrptRule",
-            "-Namespace",
-            ".",
-            "-Comment",
-            FZ_MAGIC,
-            "-NameServers",
-            &dns_config_string,
-        ])
-        .status()?;
+    let dns_config_string = dns_config
+        .iter()
+        .map(|ip| ip.to_string())
+        .collect::<Vec<_>>()
+        .join(";");
+
+    let hkcu = winreg::RegKey::predef(winreg::enums::HKEY_LOCAL_MACHINE);
+    let base = Path::new("SYSTEM")
+        .join("CurrentControlSet")
+        .join("Services")
+        .join("Dnscache")
+        .join("Parameters")
+        .join("DnsPolicyConfig")
+        .join(NRPT_REG_KEY);
+
+    let (key, _) = hkcu.create_subkey(base)?;
+    key.set_value("Comment", &FZ_MAGIC)?;
+    key.set_value("ConfigOptions", &0x8u32)?;
+    key.set_value("DisplayName", &"Firezone SplitDNS")?;
+    key.set_value("GenericDNSServers", &dns_config_string)?;
+    key.set_value("IPSECCARestriction", &"")?;
+    key.set_value("Name", &vec!["."])?;
+    key.set_value("Version", &0x2u32)?;
+
     Ok(())
 }
 
 // Must be `sync` so we can call it from `Drop`
-pub(crate) fn deactivate() -> Result<()> {
+fn deactivate() -> Result<()> {
```

Comment:
We can't delete it by just removing the registry key?

Reply:
Probably, but for backwards compatibility with older Clients we need to delete it the old way at least once per system.

We could write a file or something saying, "This system is upgraded, always use the fast version now."

The older Powershell method, IIRC, results in a random UUID for the rule, so we can't control it or delete it by its ID. We just delete everything that says "Firezone" basically.

And NRPT rules survive reboots, so if we didn't do the slow method at least once when the computer boots, it will stick around.

But this slow deactivation only runs when the system boots, after that we use the fast in-place update for `set_dns` and when the GUI signs out or closes, the deactivation latency gets hidden there.